### ISSUE TITLE
Suggestion/discussion: modularize oF

### ISSUE THREAD

**Issue by thomasgeissl** [openframeworks/openFrameworks#4929]
Hello,
I saw you are restructuring oF quite a bit. I'd like to suggest to make oF more modular. Right now it is one bit framework and I use it for lot of different programming jobs. But lets say I am building a command line tool or a simple server, then I like to keep it as minimal as possible, without graphics and audio stuff. 
Ideally I could just use the modules, which are really needed. Maybe we could split it into modules/addons like core, audio, graphics, network, ...

Also things like ofSerial and the firmata classes could live as an addon. I think ofxOsc is used more often than ofSerial and it is an addon. Do other classes depend on ofSerial? I don't see why one is an addon and the other one part of the core.

Here are some advantages I see:

* Apps would be less error prone. 
* Probably easier to maintain, test, debug.
* oF could be used even on non supported devices, by just not using critical modules, e.g. graphics module on micro computers.
* oF modules could be used inside other frameworks, e.g. oF core inside Qt

I know this would be a big change, but I would like to know your opinions. What are the cons?

Thomas

**Comment by firmread** [comment]
I like the general idea of what you're proposing, although I don't know any implication of how to implement this :P I think it is similar to what is discussed here in https://github.com/openframeworks/openFrameworks/pull/4830. Where different templates could solve this like OF with out Fmod. Maybe more template that only have liter version of OF could satisfy your what you're thinking? 
Looking forward to hear more better responds here too :)

**Comment by arturoc** [comment]
yes this is one of the ideas that came up in the last dev conference. the only similar step we've taken until now is beginning to remove the dependency with poco and move it into an addon since it's the biggest library we use and it was always problematic with new platforms.

the way to do it would probably be to move bits of the core into addons, probably the easiest to move would be the platform specific code, for example the video elements could live in ofxGStreamer, ofxAVFoundation, ofxDirectShow. That also has the advantage that you could choose the specific implementation you want to use simply by adding one or other addon to the project.

**Comment by thomasgeissl** [comment]
Is there a protocol of the dev conference online?

**Comment by arturoc** [comment]
most of those documents have links in the development section in the web: http://localhost/development/ in particular this should be documented in the roadmap to 1.0: https://docs.google.com/document/d/1GjgjQZCHTcOD3cT9OFSl0AZ1nhfYwwvxTl5L4OJE6qA/edit

**Comment by thomasgeissl** [comment]
Thanks, I am looking forward to see this happening.
